File: lab2-functionalprimer/haskell/func_as_arg.hs
{--create inpFunc
inpFunc a b = [a..b]

-- Define applicatorFunc
applicatorFunc inpFunc a b s = 
    if s == 's' 
    then fromIntegral (sum (inpFunc a b))  -- Sum of the list
    else fromIntegral (sum (inpFunc a b)) / (b - a + 1)  -- Average of the list

main = do
    let result = applicatorFunc inpFunc 1 10 's'  -- Call applicatorFunc with inpFunc and 's' as args
    putStrLn ("sum = " ++ show result)

-}

-- make the function above using guards 
inpFunc a b = [a..b]

-- Define applicatorFunc
applicatorFunc inpFunc a b s | s=='s'    = sum (inpFunc a b)
                             | otherwise = (sum(inpFunc a b))/(b-a+1)
    

main = do
    let result = applicatorFunc inpFunc 1 5 's'  -- Call applicatorFunc with inpFunc and 's' as args
    putStrLn ("sum = " ++ show result)









{-applicatorFunc :: [Int] -> Char -> Double
applicatorFunc inpFunc s = 
    if s == 's' 
        then fromIntegral (sum inpFunc) 
        else fromIntegral (sum inpFunc) / fromIntegral (length inpFunc)

main :: IO ()
main = do
    putStrLn "Enter the start of the range: "
    a <- readLn
    putStrLn "Enter the end of the range: "
    b <- readLn
    putStrLn "Enter 's' for sum or any other character for average: "
    s <- getChar
    let inpFunc = [a..b]
    let result = applicatorFunc inpFunc s
    putStrLn ("Result = " ++ show result)
-}

File: lab2-functionalprimer/python3/func_as_arg.py
"""
def listFunc():
    return [i for i in range(1, 6)] #Create list of ints from 1 to 5, Haskell equivalent [1..5]

def applicatorFunc(inpFunc, s):
    if s=='s':
        return sum(inpFunc())
    else:
        return sum(inpFunc())/5

print(applicatorFunc(listFunc, 's'))
"""

def listFunc(a, b):
    #Create a list of integers from a to b, equivalent to Haskell's [a..b].
    return list(range(a, b + 1))

def applicatorFunc(inpFunc, s, a, b):
    #Computes sum or average of numbers from a to b based on the character s.
    numbers = inpFunc(a, b)
    if s == 's':
        return sum(numbers)
    else:
        return sum(numbers) / len(numbers)

# Example usage
print(applicatorFunc(listFunc, 'a', 1, 5))  # Calculates average of numbers from 1 to 10
print(applicatorFunc(listFunc, 's', 1, 10))  # Calculates sum of numbers